### ISSUE TITLE
Add a compositional neural instance retriever to handle incompleteness in the data

### ISSUE THREAD

**Issue by Jean-KOUAGOU** [dice-group/owlapy#123]
- A compositional neural instance retriver can parse and encode a class expression into a continuous vector, which is then used together with the embedding of an individual to produce a probability "that the individual is an instance of the class expression"

- We plan to implement three compositional neural instance retrievers based on: 1) a transformer architecture, 2) an rnn architecture, 3) the NAND operator and direct interpretations of DL concetpts as sets

- An initial implementation with a transformer architecture is available at https://github.com/dice-group/CoNeuralReasoner

- Integration should start in the coming weeks

**Comment by Demirrr** [comment]
@LckyLke and @Jean-KOUAGOU  

We have two different neural OWL reasoner that can retrieve instances by using embeddings/neural networks
@LckyLke and @Jean-KOUAGOU. Given that @LckyLke 's approach already implemented/experimented etc., please create a feature branch @LckyLke and integrate your model into owlapy that is already integerated in ontolearn (later on we will depricate it in ontolearn)


Todos for @LckyLke :

- [ ] Create a python package named owl_neural_reasoners
- [ ] Create an abstract.py that specifies the interface of our neural reasoners
- [ ] Create a script that only contains your approach
- [ ] Create an example for the retrival task into examples folder ( this basically => https://github.com/dice-group/Ontolearn/blob/develop/examples/retrieval_eval.py)
- [ ] Write a regression test for the retreival performances on family dataset

Once these steps are done, @Jean-KOUAGOU can start their integration of compostional neural instance retriever

Please thumps up of you like the plan

**Comment by LckyLke** [comment]
We have to see how the if the interface, i.e. the abstract neural reasoner, needs adjustment depending on  which methods the neural reasoners should really **have** to implement.

Also now we have dicee (for KGE) and ontolearn (for example script) as dependecies

**Comment by Demirrr** [comment]
I have been thinking about following our design
```python
from owlapy.owl_ontology_manager import SyncOntologyManager, OntologyManager
from owlapy.owl_reasoner import StructuralReasoner, SyncReasoner
cwr_reasoner=StructuralReasoner(ontology=OntologyManager().load_ontology(path="KGs/Family/father.owl"))
owr_reasoner=SyncReasoner(ontology=SyncOntologyManager().load_ontology(path="KGs/Family/father.owl"),reasoner="HermiT")
neural_reasoner_A=NeuralReasonerA(ontology=NeuralReasonerOntologyManager(), ...)
neural_reasoner_B=NeuralReasonerB(ontology=NeuralReasonerOntologyManager(), ...)

```

**Comment by Demirrr** [comment]
Why do you think that we need ontolearn here ?

**Comment by LckyLke** [comment]
> Why do you think that we need ontolearn here ?

Yes, I guess I should refactor and use Structural reasoner directly for the reference results

**Comment by LckyLke** [comment]
Now we dont use ontolearn anymore :)

**Comment by Demirrr** [comment]
You are right that we do not want to load an ontology into memory for KGE base models if the path of a pretrained model is given.


The purpose of `NeuralReasonerOntologyManager()` is to encapsulate the preprocessing steps, read and write computations within an ontology manager class. By doing so, we can abide by the interface of using reasoners (e.g. StructuralReasoner and SyncReasoner)

**Comment by LckyLke** [comment]
I will think of a sensible implementation for this class as well than 👍🏼

**Comment by LckyLke** [comment]
> @LckyLke and @Jean-KOUAGOU
> 
> We have two different neural OWL reasoner that can retrieve instances by using embeddings/neural networks @LckyLke and @Jean-KOUAGOU. Given that @LckyLke 's approach already implemented/experimented etc., please create a feature branch @LckyLke and integrate your model into owlapy that is already integerated in ontolearn (later on we will depricate it in ontolearn)
> 
> Todos for @LckyLke :
> 
>     * [x]  Create a python package named owl_neural_reasoners
> 
>     * [x]  Create an abstract.py that specifies the interface of our neural reasoners
> 
>     * [x]  Create a script that only contains your approach
> 
>     * [x]  Create an example for the retrival task into examples folder ( this basically => https://github.com/dice-group/Ontolearn/blob/develop/examples/retrieval_eval.py)
> 
>     * [x]  Write a regression test for the retreival performances on family dataset
> 
> 
> Once these steps are done, @Jean-KOUAGOU can start their integration of compostional neural instance retriever
> 
> Please thumps up of you like the plan

for the regression test to work with the github action we should provide a pretrained model for Family somewhere to fetch during testing.

**Comment by Demirrr** [comment]
Why don't we train the model from scratch and evaluate it as we have been doing in the dice-embeddings ( see for example https://github.com/dice-group/dice-embeddings/blob/develop/tests/test_regression_conex.py)

**Comment by LckyLke** [comment]
What hardware are we running this action on? I thought this might take too long but if there is gpu available this would work ofc 👍🏻

**Comment by Demirrr** [comment]
It doesn't take much time for the time even without gpu provided that the number of epochs, embedding dims are set accordingly

**Comment by LckyLke** [comment]
Ok I will test how long it takes :)

**Comment by LckyLke** [comment]
> You are right that we do not want to load an ontology into memory for KGE base models if the path of a pretrained model is given.
> 
> The purpose of `NeuralReasonerOntologyManager()` is to encapsulate the preprocessing steps, read and write computations within an ontology manager class. By doing so, we can abide by the interface of using reasoners (e.g. StructuralReasoner and SyncReasoner)

Should preprocessing involve training a model if a kb path is provided?

**Comment by Demirrr** [comment]
I would say no because the training process might differ between approaches.

**Comment by LckyLke** [comment]
Cause of Issue #124 I removed the neural ontology manager for the time being

**Comment by Demirrr** [comment]
Closing due to inactivity (no updates since 2024-12-12) and no linked implementation work in this repo. The transformer-based prototype referenced (CoNeuralReasoner) can be tracked/integrated via a fresh issue when work actually resumes.
